Continueing debugging TOO MANY API CALLS

Zayen-X opened the pull request:
Done:
- `web3context.conntected` is the definitive source of truth for "is it a connected wallet"
- Separating out load of app vs account. 

TODO:
- [ ]  Still a swam of errors in the console when attempting to connect to providers.

Zayen-X committed:
moving connect call lower to make it  the definitive source of 'isConnected'.

Zayen-X commented:
Adding @appleseed-iii 's comments here for the record:
> I think there are two polling issues. I obviously could be wrong so take that as a disclaimer:

> 1. The current infura & alchemy api credentials are overloaded & we get lots of 429 errors as a result. Somewhere in ethers the failing backend network calls are resulting in polling to check if the network connection is repaired (it's not considering our case where the network connection issue isn't temporary). This can be tested/proven by replacing the production dapp's infura / alchemy api keys with clean keys. 

> 2. So when we fix #1 above we eliminate one source of polling. Then the second source of polling is the polling for the web3Provider network change. For some reason the way our code works (I don't know why yet) that network change polling also runs on JsonRpcProviders. This can be proven by simply commenting out lines 132-135 of web3Context & watching the network tab in the chrome inspector (only if you have fixed #1 above also). The lines I highlighted below: [`provider.on("network")` in `web3Context.tsx#_initProviders()`]

> I think the long-term answer is to separate the way we initiate providers. There is some sort of code overlap where JsonRpcProviders are getting those listeners

wKich commented:
I have two proposals to fix this issue:
- Write a simple queue, each API call will be sent to that queue. The queue processor takes calls from the queue and executes them with a limit strategy, so we can't hit 429 errors. The downside of this way is it can increase time of loading app, but I think it might be faster than now when we receive a lot of 429 errors.
- Another way a little complicated. I guess that we fetch a lot of data at once even if we don't need some of it. For example, on main dashboard page, we need to show only bonds profits, price, circulation/total supply and something else, so we don't need to get ohm balance, allowance and other stuff. So we need to make data in our store available on-demand and then we can make only those requests that are only needed.

These two approaches are workaround. Ideally we need to deploy our API backend service, there we can cache data and significantly reduce call to infura or other providers

appleseed-iii commented:
@wKich 
For your point `#1` IMO I'd rather just put together a more permanent solution (like the API backend service). 
For your point `#2`, yes, I can confirm what you are saying. For sure (in this branch) the app currently makes about 65 API requests every time the app is loaded or refreshed. We don't limit calls to specific pages (staking, bonding, etc). We just get everything all the time. 

I think the immediate roadmap to fix should be:

1. Fix the network polling issue(the 2nd item) in @lolchocotaco's comment above. I'm working on this now.
2. Push this branch to production, which should cut the API requests in half (65 per refresh, rather than 130 plus).
3. Your Item 2 above, limit api calls to where they are needed in the app. I haven't looked to see how many calls this could reduce in theory.

This roadmap should be pretty quick to implement/deploy unless I'm overlooking some complexity in there.

wKich commented:
@appleseed-iii I think the 1st option is pretty simple to implement, but the 2nd one requires to make some refactor I can't say how effective it could be. But we can combine these options. I can make a PR with the 1st one and take a look how it helps. In the end we definitely need to deploy api service

appleseed-iii commented:
Pr https://github.com/OlympusDAO/olympus-frontend/pull/366 fixes the network polling issue, eliminating the set of backend calls that would run every 4 seconds. Please merge that pr into this one before merging this onto develop branch. I'll take a look at the 429 error retries tomorrow.

appleseed-iii commented:
FYI, for the 429 throttling, [ethers has some throttling implementation](https://github.com/ethers-io/ethers.js/commit/88c7eaed061ae9a6798733a97e4e87011d36b8e7) but it doesn't appear that it's working for us bc I don't ever see [showThrottleMessage](https://github.com/ethers-io/ethers.js/commit/88c7eaed061ae9a6798733a97e4e87011d36b8e7#diff-9de0b6aaa58d0ca93a08f6e1a532023d397e517fffb3f23658fcbb8219f102e2R460) running. (upon further review the ethers throttling implementation is only for the community api keys).

@lolchocotaco I'm also wondering if trying to prevent the 429 retries is best practice. We're rate-limited routinely right now & the end result is the app not loading properly for users. By introducing retry-cancellation logic the end-result will still be the same for the user & meanwhile has the potential to introduce other unintended side effects?

Alternatively, we could just set up some clean API keys for Infura & Alchemy & hopefully eliminate the rate limiting moving forward (as a result of the other changes in this pull request). Every time we're approaching Rate Limits we could add another API key to the "intelligent" load balancer.

appleseed-iii committed:
Rework provider initialization  (#366)

* rework provider initialization so that listeners only apply to wallet providers where needed, eliminating network polling
move _initListeners into a useEffect hook that is triggered by walletProvider state change. In previous implementations _initListeners was rplaced in the code *after* setProvider, but state is not immediately changed in react and so _initListeners was actually running *before* the provider state was updated.

* refactoring so that we only have ONE provider state variable. It changes from JsonRpc on initialization to Web3Provider if the user has connected wallet. TMeanwhile, _initListeners only runs for Web3Providers now, rather than for JsonRpcProviders as was the case before this branch

Zayen-X committed:
limit to one alchemyAPI and fix app reload on 429s

Zayen-X committed:
comments

Zayen-X committed:
reverting function signature, idk why we use loadProvider

appleseed-iii commented:
I tested your last three commits. Everything looks good on my end.
